src/PsrLogAdapter.php

```diff
@@ -0,0 +1,72 @@
+<?php
+/*
+ * Copyright 2023-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB;
+
+use MongoDB\Driver\Logging\Logger as DriverLoggerInterface;
+use Psr\Log\LoggerInterface as PsrLoggerInterface;
+use Psr\Log\LogLevel as PsrLogLevel;
+
+class PsrLogAdapter implements DriverLoggerInterface
+{
+    private PsrLoggerInterface $psrLogger;
+
+    public function __construct(PsrLoggerInterface $psrLogger)
+    {
+        $this->psrLogger = $psrLogger;
+    }
+
+    public function log(int $level, string $domain, string $message): void
+    {
+        $this->psrLogger->log(
+            $this->driverLevelToPsrLevel($level),
+            $domain . ': ' . $message,
+        );
+    }
+
+    // TODO: Refactor this method as a private const array map on PsrLogAdapter
+    private function driverLevelToPsrLevel(int $level): string
+    {
+        switch ($level) {
+            /* libmongoc considers "critical" less severe than "error" so route
+             * both levels to "error" in the PSR logger. */
+            case DriverLoggerInterface::LEVEL_ERROR:
+            case DriverLoggerInterface::LEVEL_CRITICAL:
+                return PsrLogLevel::ERROR;
+
+            case DriverLoggerInterface::LEVEL_WARNING:
+                return PsrLogLevel::WARNING;
+
+            case DriverLoggerInterface::LEVEL_MESSAGE:
+                return PsrLogLevel::NOTICE;
+
+            case DriverLoggerInterface::LEVEL_INFO:
+                return PsrLogLevel::INFO;
+
+            /* PSR does not define a "trace" level, so route both levels to
+             * "debug". That said, trace logging is very verbose and should not
+             * be encouraged when using a PSR logger. */
+            case DriverLoggerInterface::LEVEL_DEBUG:
+            case DriverLoggerInterface::LEVEL_TRACE:
+                return PsrLogLevel::DEBUG;
+
+            /* A default case is intentionally not handled since libmongoc and
+             * PHPC should avoid emitting bogus levels; however, we can consider
+             * throwing an UnexpectedValueException here. */
```

Comment:
Throwing may be undesirable, but we may want to emit a warning or error about an undefined level and logging the message with the same level for good measure

Reply:
This is the easiest place to log a message. I agree with @alcaeus that you should log a warning that the level is incorrect.

Reply:
> emit a warning or error about an undefined level

Can you clarify what you mean by "emit a warning or error"? Are you suggesting we log an additional message through the PSR logger since we already have access to it?

> logging the message with the same level for good measure

And here, are you suggesting we log the original message (received from PHPC) at a warning/error level?

What are you thoughts on logging a single message that reports the invalid level, and appends the original log message? Then we only end up with a single log line with all of the original information _and_ our error.

Reply:
We now throw a UnexpectedValueException that includes the original level, domain, and message.